packages/flutter/test/widgets/system_context_menu_test.dart

```diff
@@ -895,4 +895,121 @@ void main() {
+        );
+      });
+
+      final List<IOSSystemContextMenuItem> items = <IOSSystemContextMenuItem>[
+        const IOSSystemContextMenuItemCopy(),
+        IOSSystemContextMenuItemCustom(
+          title: 'Custom Action 1',
+          onPressed: () {
+            customAction1Called = true;
+          },
+        ),
+        IOSSystemContextMenuItemCustom(
+          title: 'Custom Action 2',
+          onPressed: () {
+            customAction2Called = true;
+          },
+        ),
+      ];
+
+      final TextEditingController controller = TextEditingController(text: 'test text');
+      addTearDown(controller.dispose);
+      
+      await tester.pumpWidget(
+        Builder(
+          builder: (BuildContext context) {
+            final MediaQueryData mediaQueryData = MediaQuery.of(context);
+            return MediaQuery(
+              data: mediaQueryData.copyWith(supportsShowingSystemContextMenu: true),
+              child: MaterialApp(
+                home: Scaffold(
+                  body: Center(
+                    child: TextField(
+                      controller: controller,
+                      contextMenuBuilder:
+                          (BuildContext context, EditableTextState editableTextState) {
+                            return SystemContextMenu.editableText(
+                              editableTextState: editableTextState,
+                              items: items,
+                            );
+                          },
+                    ),
+                  ),
+                ),
+              ),
+            );
+          },
+        ),
+      );
+
+      await tester.tap(find.byType(TextField));
+      final EditableTextState state = tester.state<EditableTextState>(find.byType(EditableText));
+      expect(state.showToolbar(), true);
+      await tester.pump();
+
+      expect(find.byType(SystemContextMenu), findsOneWidget);
+      expect(itemsReceived, hasLength(1));
+      expect(itemsReceived.last, hasLength(3));
+      
+      expect(itemsReceived.last[0], equals(const IOSSystemContextMenuItemDataCopy()));
+      expect(itemsReceived.last[1], isA<IOSSystemContextMenuItemDataCustom>());
+      expect((itemsReceived.last[1] as IOSSystemContextMenuItemDataCustom).title, 'Custom Action 1');
+      expect(itemsReceived.last[2], isA<IOSSystemContextMenuItemDataCustom>());
+      expect((itemsReceived.last[2] as IOSSystemContextMenuItemDataCustom).title, 'Custom Action 2');
+      
+      final SystemContextMenu contextMenu = tester.widget<SystemContextMenu>(
+        find.byType(SystemContextMenu),
+      );
+      final SystemContextMenuController controller2 = contextMenu.controller;
+      
+      final IOSSystemContextMenuItemCustom customItem1 = items[1] as IOSSystemContextMenuItemCustom;
+      final IOSSystemContextMenuItemCustom customItem2 = items[2] as IOSSystemContextMenuItemCustom;
+      
+      controller2.handleCustomContextMenuAction(customItem1.hashCode.toString());
```

Comment:
Can you also test that a call from onPerformCustomAction works? You can simulate the platform message using [TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger.handlePlatformMessage](https://main-api.flutter.dev/flutter/flutter_test/TestDefaultBinaryMessenger/handlePlatformMessage.html).

No need to use handlePlatformMessage every time, it's also fine to use handleCustomContextMenuAction directly like you are here. Just make sure that the onPerformCustomAction gets tested at least once too.

Reply:
Also try the error state: send that message when there is no _systemContextMenuClient.